log/std/logger.go

```diff
@@ -0,0 +1,224 @@
+// Package std is the implementation of the logger interface with the standard log package.
+package std
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"sort"
+	"strings"
+
+	patronLog "github.com/beatlabs/patron/log"
+)
+
+var levelMap = map[patronLog.Level]string{
+	patronLog.DebugLevel: "DBG",
+	patronLog.InfoLevel:  "INF",
+	patronLog.WarnLevel:  "WRN",
+	patronLog.ErrorLevel: "ERR",
+	patronLog.FatalLevel: "FTL",
+	patronLog.PanicLevel: "PNC",
+}
+
+// Logger implementation of the std log.
+type Logger struct {
+	level      patronLog.Level
+	fields     map[string]interface{}
+	fieldsLine string
+	out        io.Writer
+	debug      *log.Logger
+	info       *log.Logger
+	warn       *log.Logger
+	error      *log.Logger
+	panic      *log.Logger
+	fatal      *log.Logger
+}
+
+// New constructor.
+func New(out io.Writer, lvl patronLog.Level, fields map[string]interface{}) *Logger {
+	fieldsLine := createFieldsLine(fields)
+
+	return &Logger{
+		out:        out,
+		debug:      createLogger(out, patronLog.DebugLevel, fieldsLine),
+		info:       createLogger(out, patronLog.InfoLevel, fieldsLine),
+		warn:       createLogger(out, patronLog.WarnLevel, fieldsLine),
+		error:      createLogger(out, patronLog.ErrorLevel, fieldsLine),
+		panic:      createLogger(out, patronLog.PanicLevel, fieldsLine),
+		fatal:      createLogger(out, patronLog.FatalLevel, fieldsLine),
+		level:      lvl,
+		fields:     fields,
+		fieldsLine: fieldsLine,
+	}
+}
+
+func createFieldsLine(fields map[string]interface{}) string {
+	if len(fields) == 0 {
+		return ""
+	}
+
+	// always return the fields in the same order
+	keys := make([]string, 0, len(fields))
+	for key := range fields {
+		keys = append(keys, key)
+	}
+
+	sort.Strings(keys)
+
+	sb := strings.Builder{}
+
+	for _, key := range keys {
+		sb.WriteString(key)
+		sb.WriteRune('=')
+		sb.WriteString(fmt.Sprintf("%v", fields[key]))
+		sb.WriteRune(' ')
+	}
+
+	return sb.String()
+}
+
+func createLogger(out io.Writer, lvl patronLog.Level, fieldLine string) *log.Logger {
+	logger := log.New(out, levelMap[lvl]+" "+fieldLine, log.LstdFlags|log.Lmicroseconds|log.LUTC|log.Lmsgprefix|log.Lshortfile)
+	return logger
+}
+
+// Sub returns a sub logger with additional fields.
+func (l *Logger) Sub(fields map[string]interface{}) patronLog.Logger {
+	for key, value := range l.fields {
+		fields[key] = value
+	}
+
+	return New(l.out, l.level, fields)
+}
+
+// Fatal logging.
+func (l *Logger) Fatal(args ...interface{}) {
+	if !l.shouldLog(patronLog.FatalLevel) {
+		return
+	}
+
+	output(l.fatal, args...)
+	os.Exit(1)
+}
+
+// Fatalf logging.
+func (l *Logger) Fatalf(msg string, args ...interface{}) {
+	if !l.shouldLog(patronLog.FatalLevel) {
```

Comment:
Do I understand correctly that with this line, we disallow loggers that don't have a high-enough level to fatal or panic the application?

Reply:
This line checks if the loggers level is appropriate to allow logging.
Example: if the logger is setup to allow Info and we log a debug message it should not log.